### PR TITLE
docker: add support for CUDA in docker

### PR DESCRIPTION
**Pull request by canardleteer**
Assuming one has the [nvidia-container-toolkit](https://github.com/NVIDIA/nvidia-container-toolkit) installed on Linux, or is using a GPU enabled cloud, `cuBLAS` should be accessible inside the container.

I'm not very familiar with Github Actions, nor the available execution environments available on GitHub. I wouldn't suggest pre-building these and putting them in the registry, unless there's a CI path for them.

I'm not too sure what the right path for putting these into a registry is. But I did want to contribute so people could try it locally!

### CODE REVIEW

**Comment by canardleteer** [comment]
I should add a note, I haven't been able to  get the:
```
-p "Building a website can be done in 10 simple steps:"
```

...part of the examples in `README.md` to work for me in Linux because of escaping. I have tested it with `--random-prompt` however and it works.

I just wanted to keep it in line with the other Docker examples.

File: README.md
Comment (JohannesGaessler) [inline README.md:536]: A comment that explains the difference between these would be useful.

**Review by JohannesGaessler — COMMENTED** [review]
I was thinking something like this would be nice to have but ultimately prioritized fixing the  current issues like memory management first. Thanks.

File: README.md
Comment (canardleteer) [reply README.md]: Added a note describing them, as well renamed their labels to be `cuda-*`.

File: README.md
Comment (JohannesGaessler) [inline README.md:557]: You forgot to rename the containers here.
Also I get the following error when I try to run the containers:
```
/home/johannesg/Projects/llama.cpp [git::feat/docker-cuda *] [johannesg@johannes-pc] [17:45]
> docker run --gpus all -v /mnt/Windows10/Users/Johannes/Projects:/models local/llama.cpp:full-cuda --run -m /models/llama-7b-ggml-q4_0.bin -p "Building a website can be done in 10 simple steps:" -n 512 --n-gpu-layers 99
error: unknown argument: a
usage: ./main [options]
```

**Review by JohannesGaessler — CHANGES_REQUESTED** [review]

File: README.md
Comment (JohannesGaessler) [reply README.md:557]:
```
/home/johannesg/Projects/llama.cpp [git::feat/docker-cuda *] [johannesg@johannes-pc] [18:04]
> docker run --gpus all -v /mnt/Windows10/Users/Johannes/Projects:/models local/llama.cpp:light-cuda --run -m /models/llama-7b-ggml-q4_0.bin -p "Building a website can be done in 10 simple steps:" -n 512 --n-gpu-layers 99
error: unknown argument: --run
usage: /main [options]
```

Comment (canardleteer) [reply README.md]: @JohannesGaessler - Good catch on the label rename, fixed.

With regards to the second issue, I've left this [comment here](https://github.com/ggerganov/llama.cpp/pull/1461#issuecomment-1547050032), and there's also an [open issue here](https://github.com/ggerganov/llama.cpp/issues/1463) on the subject.

I mirrored what was already in the `README.md` so that issue can be resolved separately, as the problem exists in the non-CUDA instructions as well. I aimed for alignment with the other documentation, but validated with `--random-prompt` instead of `-p "something something"`.

I hope a fix for the Docker documentation can be found, that preserves the brevity of the Docker examples and doesn't turn into a treatise on bash & docker string escaping :) ... It's something else I want to look into fixing when I have time, but separately from this functionality.

Comment (canardleteer) [reply README.md]: @JohannesGaessler - In your second failed docker command:

It appears you're attempting to run the `full` arguments on the `light` container.

```
> docker run --gpus all -v /mnt/Windows10/Users/Johannes/Projects:/models local/llama.cpp:light-cuda --run -m /models/llama-7b-ggml-q4_0.bin -p "Building a website can be done in 10 simple steps:" -n 512 --n-gpu-layers 99
error: unknown argument: --run
```

The `--run` argument is not available in either the `light` or `light-cuda` examples. This behavior matches the original docker image as well.

Comment (JohannesGaessler) [reply README.md]: Thank you for the hint regarding the second command. It now works. What I've noticed is that CTRL+C cannot be used to interrupt the token generation (but I assume that is then an issue on master as well?). In any case, I got both containers to work with GPU acceleration and there don't seem to be issues that don't exist on master then.

**Review by JohannesGaessler — APPROVED** [review]
I'm not particularly knowledgeable about docker but to me it looks like everything is in order now.

**Comment by canardleteer** [comment]
Fixed a trailing white space error from the `Makefile` (and installed `editorconfig` for the future).

File: README.md
Comment (canardleteer) [reply README.md]: I run into that issue with ctrl+c in Docker as well, regardless of if it's the CUDA or non-CUDA version.

It does however work, when I'm in interactive mode and I press it twice. I don't know what the root cause of it is yet, but when I get a chance, I plan to look into it, because killing the container manually is a hassle.

Comment (JohannesGaessler) [reply README.md]: Intuitively I would have thought that starting a shell inside the docker container and then executing `./main` as you would normally would fix CTRL+C behavior. CLI arg escaping probably too. Would only be good for CLI use though.

**Comment by canardleteer** [comment]
Could I get another run of CI on this?

File: README.md
Comment (dkarlovi) [inline README.md:536]:
```suggestion
docker build -t local/llama.cpp:main-cuda -f .devops/main-cuda.Dockerfile .
```

File: README.md
Comment (dkarlovi) [inline README.md:549]:
```suggestion
2. `local/llama.cpp:main-cuda`: This image only includes the main executable file.
```

File: README.md
Comment (dkarlovi) [inline README.md:557]:
```suggestion
docker run --gpus all -v /path/to/models:/models local/llama.cpp:main-cuda -m /models/7B/ggml-model-q4_0.bin -p "Building a website can be done in 10 simple steps:" -n 512 --n-gpu-layers 1
```

**Review by dkarlovi — CHANGES_REQUESTED** [review]
Small changes to make it slightly less confusing, otherwise LGTM, it works.

**Comment by deep-pipeline** [comment]
This looks really useful for integration/deployment of llama.cpp into docker hosted services on cloud! 

Is there anything (apart from the vast amount of other activity ;-) ) holding up this being merged now? 

Um, on which note, I see that total open **pull-requests**, not issues(!), has risen from, iirc, 53 to 65 since I last checked in.. which is kinda great but also kinda scary because of implication for level of stress on supervision bandwidth..

File: README.md
Comment (canardleteer) [reply README.md:536]: In general, I'm looking to keep it in-line with the `docker` documentation upstream. This is true of the examples and the image names.

I think those things should be changed everywhere, but either upstream first and then with this change, or in both places if this lands first.

**Comment by canardleteer** [comment]
The upstream `Makefile` now has a conflict that I will need to resolve, and will do so when I get a chance.

**Comment by JohannesGaessler** [comment]
I didn't merge this PR because I wanted someone else to check it as well; as I said, I'm not very knowledgeable about Docker.

**Comment by canardleteer** [comment]
I have rebased on the latest changes in `master`. A second set of eyes on my changes + a pipeline run would alleviate my concerns about those changes :)

**Review by ggerganov — APPROVED** [review]
Will wait for an extra approve before merging as I'm unfamiliar with Docker

**Comment by deep-pipeline** [comment]
Hi folks, just a very gentle nudge again on this front - last time I checked here the total open pull-requests on this repo were about 65, a month later they are over 80.. as before I'm worried that more project approval/supervision bandwidth needs to be allocated to clear backlog a bit so older 'finished' PRs like this don't age stuck in limbo, suffer new integration issues and need revisiting..

It would be a shame if good work gets buried/lost/stuck with PRs not folded into main (and if sprawling PR numbers mean stale PRs are not being explicitly closed or duplicates being coalesced).

The docker support stuff ought to be relatively independent of other changes - maybe worth a final CI check and folding in this one now - any problems for nvidia-docker use can be sorted via bug-reports by people trying it (who currently don't have the easy option). 

Best, M.

**Pull request merged**
